Expand finite volume support

lindsayad opened the pull request:
Includes:
- Gradient reconstruction
- Non-orthogonal corrections
- Auxiliary finite volume variable support
- Verification of FV for advection-diffusion-reaction type problems

Remaining to-do:

- [x] Verify that "non-local" computation of gradient works with block restriction

Closes #15063 #15836 #15066 #15825 #15329

Refs #15062, e.g. this adds support for coupling FV into FE

lindsayad committed:
All framework level changes to support INSFV

Includes:
- Gradient reconstruction
- Non-orthogonal corrections
- Auxiliary finite volume variable support
- Verification of FV for advection-diffusion-reaction type problems

lindsayad commented:
@rwcarlsen With 115 commits in the INSFV branch it was a little time consuming extricating only the framework commits that would support gradient reconstruction testing. So here you go in one big squashed commit all the framework level changes from INSFV support. I was actually pleased to see that while this is 5k lines, there's actually not nearly that much in terms of framework changes. Most of the lines are from testing

lindsayad committed:
Add option for restricting tests based on indexing

lindsayad committed:
Restrict tests relying on global indexing

rwcarlsen reviewed (changes requested):
Here is my first pass through.  Mostly questions and minor nitpicks.  Could you provide maybe some suggestions of themes or bigger things/changes for me to focus down on for the review?  That would help make a second pass more effective.

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1132,14 +985,35 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
   ///@{ accessors for the FaceInfo objects
   unsigned int nFace() const { return _face_info.size(); }
-  std::vector<FaceInfo> & faceInfo()
+  std::vector<FaceInfo *> & faceInfo()
```
Because of the addition of `_all_face_info`

lindsayad replied:
I think I'm going to actually convert this to a container of `const FaceInfo *` and encapsulate any modification of `FaceInfo` in the owning `MooseMesh`.

lindsayad commented:
> Could you provide maybe some suggestions of themes or bigger things/changes for me to focus down on for the review?

I think the most meaningful code to look over are the methods added at the end of `MooseVariableFV`. Another design question is whether `FVFaceInterface` should exist at all. My vote would be to move everything into `FVUtils` which is what I was gradually doing over the course of the INSFV work. I was thinking about whether to create an `FV` namespace nested inside of the `Moose` namespace. What are your thoughts on that? I could go either way.

rwcarlsen commented:
I'm all for using composition and functional style functioning functions as much as possible - which is a fancy way of saying "yes - I also would prefer making the fvfaceinterface functions just global/helper functions rather than having a separate interface everyone inherits from".  An FV namespace sounds great.  It's always much much easier to remove or rename a namespace than to introduce one later.

lindsayad committed:
Documentation of new FaceInfo-based methods

lindsayad committed:
Deprecate *Contribution* methods

These methods were confusing because they fed into different data
members in `Assembly`. I don't see any reason for the existence of two
sets of these cached Jacobian data members, so I eliminated one set, and
then I deprecated all methods with "contribution" in the name because to
me it just propagates the confusion. Much better to have multiple
overloads with the same clear name IMO

lindsayad committed:
Fix deprecated master_boundary inputs

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1132,14 +985,35 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
   // const
   ///@}
 
+  void computeFaceInfoFaceCoords(const SubProblem & subproblem);
```
Done (hopefully)

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1308,7 +1182,11 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
   /// FaceInfo object storing information for face based loops
-  std::vector<FaceInfo> _face_info;
+  std::vector<FaceInfo *> _face_info;
+  std::vector<FaceInfo> _all_face_info;
```
Done (hopefully)

lindsayad commented on framework/include/problems/DisplacedProblem.h:
```diff
@@ -268,6 +269,7 @@ class DisplacedProblem : public SubProblem
   virtual void cacheJacobianNonlocal(THREAD_ID tid);
   virtual void cacheJacobianNeighbor(THREAD_ID tid) override;
   virtual void addCachedJacobian(THREAD_ID tid) override;
+  virtual void addCachedJacobianContributions(THREAD_ID tid) override;
```
So what I've done is eliminate the `cached_jacobian_contributions*` members in `Assembly` and I've deprecated methods with "Contribution" in the name because it is very confusing to me.

lindsayad commented on framework/include/variables/MooseVariableField.h:
```diff
@@ -83,7 +83,13 @@ class MooseVariableField : public MooseVariableFieldBase
+   * Degree of freedom value setters
+   */
   virtual void setDofValue(const OutputData & value, unsigned int index) = 0;
+  virtual void setElementalValue(const OutputType & value) = 0;
```
Ok I was able to eliminate this API by adding a private `setDofValueHelper` to `AuxKernel`.

lindsayad committed:
Private AuxKernel setter helper instead of whole new variable API

lindsayad committed:
Remove FVFaceInterface

lindsayad commented:
Ugh, I think this is going to need distributed mesh fixes included in #15338 and #15655

rwcarlsen commented:
ouch.  - so we have a triple stack - of libmesh, global dofs, this.  And it will be a quadruple stack when I start pushing up commits for fv array vars.

lindsayad committed:
Don't ask for neighbor information when its remote

rwcarlsen commented:
When you get a chance - let's get this rebased and conflicts resolved.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit b51a1fac743ce11a1e65018f25a478241c24ed44
+Subproject commit 16f1801381b3a63b644002330b3d029eed35ff3b
```
**Caution!** This contains a submodule update

lindsayad commented:
Hammering away on tests in #15338. Once those are passing, I can ensure that tests on this PR are working

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit b51a1fac743ce11a1e65018f25a478241c24ed44
+Subproject commit 3f9015493934a777ff9228616e613144a664de7c
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit b51a1fac743ce11a1e65018f25a478241c24ed44
+Subproject commit ef2cacfd227f75f812f163dad2e4c1b61e1e75e1
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/588942/) on 3296a98 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15825/site)

This comment will be updated on new commits.

lindsayad committed:
Qualify with this-> because poor gcc 5 is confused

lindsayad committed:
Add documentation

lindsayad committed:
Rip out FVFunctionDiffusionBC

After writing documentation for the other new classes, I realized that
unless I've forgotten it, there's not a good reason for having this
class. Perhaps when I get to merging the INSFV work, a reason will
manifest itself, in which case I will revert this commit. Until then
though, don't write code you don't need. Keep it simple, stupid

lindsayad committed:
Ensure that Assembly subdomain IDs are properly set

lindsayad committed:
Avoid indexing out of bounds by returning if we don't have any dofs

lindsayad committed:
Remove seemingly unused FEProblemBase::getInterfaceMaterial method

lindsayad committed:
Ensure that we call appropriate subdomainSetup method when our elem or neighbor subdomain changes

lindsayad committed:
Add neighboring subdomain dependencies to needed vars and props

lindsayad committed:
Get block restriction working in serial

Refs #15894

lindsayad committed:
Get block restriction working in parallel

I was getting lucky before on a single proc...I was getting lagged
values for material properties on the interfaces, e.g. I wasn't actually
evaluating ghost materials on the interface. I was relying on cached
values from previous evaluation of materials off the interface.
Obviously that would fail if properties were spatially dependent.
Moreover, it failed in parallel. Now we make sure that we evaluate ghost
materials on an interface, e.g. face materials from a neighbor subdomain
when doing `_fe_problem.reinitMaterialsFace` if those face materials
aren't defined on the elem subdomain, and neighbor materials from the
elem subdomain when doing `_fe_problem.reinitMaterialsNeighbor` if those
neighbor materials aren't defined on the neighbor subdomain

lindsayad committed:
Dont reinit materials with FE coupling in ghosted contexts

This requires:
1. Doing a sanity check to make sure that a material doesn't attempt to
couple in both finite element and finite volume variables
2. In FEProblemBase::reinitMaterialsFace/Neighbor removing a material
from the list of ghosted reinit materials if it's not ghostable

rwcarlsen commented:
You've done some more work on this since I last looked.  It looks like you've got a few unchecked boxes in the description still.  Where does this stand - is it ready for review or are you still tinkering with stuff?

lindsayad commented:
Probably should migrate at least the first TO-DO to an issue. That shouldn't hold this up.

I started working on generic block restriction on top of this branch, and am close to done. We can postpone that to another follow-on PR though.

lindsayad committed:
More doc for MaterialPropertyStorage::_prop_names

lindsayad committed:
Support coupling FV into FE

lindsayad committed:
Only selectively compute materials

lindsayad committed:
Reinit requisite materials for FV directly in ComputeFVFluxThread

We create a sequence of materials to use in
`FEProblemBase::getFVMatsAndDependencies`. Then we call through these
sequences in `ComputeFVFluxThread::reinitVariables`. Also, anytime a
subdomain changes in `ComputeFVFluxThread` we do a dependency check in
non-optimized modes

lindsayad committed:
Revert unused work

This reverts changes in:
- `FEProblemBase::reinitMaterialsFace`
- `FEProblemBase::reinitMaterialsNeighbor`
- `MaterialPropertyStorage`
  - I had added an ability to map from a material property to the material
    that computed it (on a per subdomain basis)

lindsayad committed:
Successful coupling of FV into FE

This required creating a lot of abstract virtuals in
`MooseVariableField` and doing some generalization in `Assembly`, e.g.
`MooseVariableFE` -> `MooseVariableField`.

rwcarlsen commented:
I'll do a final/full review on this sometime today.

lindsayad committed:
Add material prop dependency error test

lindsayad commented:
> I started working on generic block restriction on top of this branch, and am close to done. We can postpone that to another follow-on PR though.

Do you want me to postpone this? I'm about ready to push it. Finishing my last test

rwcarlsen commented:
I'll do my review as soon as you get that pushed up.

lindsayad committed:
More block restriction tests

lindsayad commented:
WTF these tests all pass on my machine and they are valgrind clean... and they just straightup seg fault on CIVET?!

rwcarlsen commented:
Beautiful code.

lindsayad commented:
Ah this is a configuration difference...

lindsayad committed:
Support block restriction with local indexing config

ajacquey commented:
Hi Alex,

I just tested your branch and I can confirm that your changes fix #15329 
Amazing work, I can't wait to test the new capabilities.

Thanks!

lindsayad commented:
> Amazing work, I can't wait to test the new capabilities.

Awesome. Thank you for taking the time to check

lindsayad committed:
Fix gradient computation in RZ coordinates

Previously was not accounting for the fact that grad(u) and
div(identity(u)) are not the same in RZ. They are in XYZ but not in RZ

lindsayad committed:
Protect access of adUDot

rwcarlsen commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -250,6 +277,28 @@ class MooseVariableFV : public MooseVariableField<OutputType>
+   * Retrieve (or potentially compute) the uncorrected gradient on the provided face
+   * @param face The face for which to retrieve the gradient
+   */
+  const VectorValue<ADReal> & uncorrectedAdGradSln(const FaceInfo & fi) const;
```
This doc comment should explain _what_ is uncorrected (i.e. cross-diffusion correction).

rwcarlsen commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -250,6 +277,28 @@ class MooseVariableFV : public MooseVariableField<OutputType>
+  const VectorValue<ADReal> & adGradSln(const Elem * const elem) const;
+
+  /**
+   * Retrieve (or potentially compute) the gradient on the provided face
+   * @param face The face for which to retrieve the gradient
+   */
+  const VectorValue<ADReal> & adGradSln(const FaceInfo & fi) const;
```
This doc comment should explain that you are getting a cross-diffusion-corrected face gradient.

rwcarlsen commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -384,9 +529,40 @@ class MooseVariableFV : public MooseVariableField<OutputType>
+  mutable std::unordered_map<const FaceInfo *, VectorValue<ADReal>> _face_to_grad;
+
+  /// A cache for storing FV \p a coefficients on elements
+  mutable std::unordered_map<const Elem *, ADReal> _elem_to_coeff;
```
what coefficients?

rwcarlsen reviewed (changes requested):
Okay this is all I could muster today.  You wouldn't think it from the (low) number of comments, but I read through this code for almost 2 hours.

lindsayad commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -412,15 +433,138 @@ ComputeFVFluxThread<RangeType>::post()
+      mooseAssert(pr.second,
+                  "Multiple objects supply property ID "
+                      << pr.second
+                      << ". Do you have materials on neighboring subdomains that define the same "
```
I didn't like doing this either. But how do I handle this?
````
[FVFluxKernels]
  [fluid_kernel]
    type = FluidKernel
    rho = rho
    mu = mu
    block = 0
  []
[]

[Materials]
  [fluid_mat]
    type = GenericConstantMaterial
    block = 0
    prop_names = 'rho mu'
    prop_values = '1 1'
  []
  [solid_mat]
    type = GenericConstantMaterial
    prop_names = 'rho'
    prop_values = '1000'
  []
[]
```

lindsayad replied:
Moreover, even if I did, how do I tell `fluid_mat` to only compute `mu` on the interface but not `rho`?

rwcarlsen replied:
Why wouldn't you want to use the neighboring subdomain's material there?  If you didn't want to use the neighboring subdomain's material in this case, then rho represents something different in each subdomain and the user should correspondingly name it differently - like rho_fluid and rho_solid.  But when they are the same name, I think we should treat them as one continuous material prop field.

rwcarlsen replied:
Wouldn't you still need mu on the interface? - you just need to ghost it instead of using the property from the neighboring subdomain (like you can with rho).

lindsayad replied:
>  What if I have 4 subdomains, and for one material property I want a different constant generic property value on each?

and

>  then rho represents something different in each subdomain and the user should correspondingly name it differently - like rho_fluid and rho_solid

seem inconsistent.

> Why wouldn't you want to use the neighboring subdomain's material there?

Because the `fluid_kernel` is restricted to block 0. The user clearly intends to only use material properties from block 0. However if they are using Dirichlet BCs, they are going to need to ghost value for rho on the neighboring side of the interface between blocks 0 and 1. So I want to make sure that `fluid_mat` gets evaluated on the neighbor side of the interface between block 0 and 1. However, if I am also evaluating a material property `rho` in `solid_mat`, then there is a race condition there.

lindsayad replied:
> Wouldn't you still need mu on the interface? - you just need to ghost it instead of using the property from the neighboring subdomain (like you can with rho).

So you're talking about whole new code for material property evaluation that would happen outside of the `Material` class. This is probably what you were thinking when we talked on the phone. That's reasonable. *But* I still think we should make it impossible for the `fluid_kernel` to use the `solid_mat` value for `rho` because that's clearly not what the user intended (because of his/her choice for block restriction). What I've implemented here accomplishes that.

rwcarlsen replied:
Yes but what you implemented here also makes impossible a very common use case.  I'm saying expecting users to supply a different property name for regions with different physics doesn't seem unreasonable.  However, the user might not have written the kernels.  And might unwittingly use a neighbor material property when they expected ghosting.  Hmmm.  Downsides to all approaches here.  I'll need to think on it a bit.

lindsayad replied:
> However, the user might not have written the kernels.

The issue is with the suppliers right? So there definitely could be a problem if the user wants to use existing `Materials` that hard-code a property name. My solution for that would be to go update those `Materials` and make them more flexible. 

> Yes but what you implemented here also makes impossible a very common use case

I do agree. I think it's totally reasonable for a user to define 'k' on multiple blocks. It's at the core of the physical concept of a material. I'm defining a thermal conductivity. Thermal conductivity is denoted by 'k'. I should be able to define 'k' for every material in my simulation.

But in the end I'd rather make the user do more work in the input file than make it possible for the simulation to do something wrong/unexpected.

If there's a straightforward implementation that accomplishes both: 1) guaranteed safety and 2) user convenience, then I'm all for it. I haven't thought of it, but maybe you will.

Ping me on slack if you want to talk about it over the phone.

lindsayad commented on framework/src/fvbcs/FVFluxBC.C:
```diff
@@ -77,6 +82,13 @@ FVFluxBC::computeJacobian(Moose::DGJacobianType type, const ADReal & residual)
     if (!jvariable.isFV())
       continue;
 
+    if (type == Moose::ElementElement &&
+        !jvariable.activeOnSubdomain(_face_info->elemSubdomainID()))
+      continue;
+    else if (type == Moose::NeighborNeighbor &&
+             !jvariable.activeOnSubdomain(_face_info->neighborSubdomainID()))
+      continue;
```
It is a correctness issue. If the variable isn't defined on a subdomain, then `Moose::adOffset` doesn't really have any meaning. Moreover, if you don't check here, you would have to check the size of `_local_ke` before doing this:
```
    _local_ke(0, 0) = residual.derivatives()[ad_offset];
```
This was the reason for my segfaults originally on CIVET when I didn't have any on my local machines. You don't get here with global indexing. If you do, and you don't have the subdomain check then your indexing into column 0 is out of bounds

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -6900,3 +6883,86 @@ FEProblemBase::reinitNeighborFaceRef(const Elem * neighbor_elem,
+    for (std::shared_ptr<MaterialBase> face_mat : this_face_mats)
+      if (face_mat->ghostable())
+      {
+        mooseAssert(!face_mat->hasStatefulProperties(),
+                    "Finite volume materials do not currently support stateful properties.");
```
It was not explicitly broken before. It's possible that it worked; probably did.

But yea with the ghosting of materials, e.g. possibly reiniting a face material from the neighbor subdomain in the same place that I'm reiniting face materials from the elem subdomain, I don't know if I understand how to swap stateful properties in that case. Maybe it's fine. It makes my head hurt when I think about this ghosting stuff

lindsayad commented on framework/src/variables/MooseVariableFV.C:
```diff
@@ -12,20 +12,53 @@
 template <typename OutputType>
 MooseVariableFV<OutputType>::MooseVariableFV(const InputParameters & parameters)
-  : MooseVariableField<OutputType>(parameters)
+  : MooseVariableField<OutputType>(parameters),
+    _solution(this->_sys.currentSolution()),
+    _phi(_assembly.template fePhi<OutputShape>(FEType(CONSTANT, MONOMIAL))),
+    _grad_phi(_assembly.template feGradPhi<OutputShape>(FEType(CONSTANT, MONOMIAL))),
+    _phi_face(_assembly.template fePhiFace<OutputShape>(FEType(CONSTANT, MONOMIAL))),
+    _grad_phi_face(_assembly.template feGradPhiFace<OutputShape>(FEType(CONSTANT, MONOMIAL))),
+    _phi_face_neighbor(
+        _assembly.template fePhiFaceNeighbor<OutputShape>(FEType(CONSTANT, MONOMIAL))),
+    _grad_phi_face_neighbor(
+        _assembly.template feGradPhiFaceNeighbor<OutputShape>(FEType(CONSTANT, MONOMIAL))),
+    _phi_neighbor(_assembly.template fePhiNeighbor<OutputShape>(FEType(CONSTANT, MONOMIAL))),
+    _grad_phi_neighbor(
+        _assembly.template feGradPhiNeighbor<OutputShape>(FEType(CONSTANT, MONOMIAL)))
```
Because of all this stuff:
```
template <typename T>
void
Assembly::copyShapes(MooseVariableField<T> & v)
{
  phi(v).shallowCopy(v.phi());
  gradPhi(v).shallowCopy(v.gradPhi());
  if (v.computingSecond())
    secondPhi(v).shallowCopy(v.secondPhi());
}

void
Assembly::copyShapes(unsigned int var)
{
  auto & v = _sys.getVariable(_tid, var);
  if (v.fieldType() == Moose::VarFieldType::VAR_FIELD_STANDARD)
  {
    auto & v = _sys.getActualFieldVariable<Real>(_tid, var);
    copyShapes(v);
  }
  else if (v.fieldType() == Moose::VarFieldType::VAR_FIELD_ARRAY)
  {
    auto & v = _sys.getActualFieldVariable<RealEigenVector>(_tid, var);
    copyShapes(v);
  }
  else if (v.fieldType() == Moose::VarFieldType::VAR_FIELD_VECTOR)
  {
    auto & v = _sys.getActualFieldVariable<RealVectorValue>(_tid, var);
    copyShapes(v);
    if (v.computingCurl())
      curlPhi(v).shallowCopy(v.curlPhi());
  }
  else
    mooseError("Unsupported variable field type!");
}
```
Since I'm adding support for coupling FV variables into FE, then I need all these methods to be implemented and not error at run-time when called

lindsayad commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -365,7 +414,103 @@ class MooseVariableFV : public MooseVariableField<OutputType>
+   * dof index
+   * @param elem The element to retrieive the solution value for
+   */
+  ADReal getElemValue(const Elem * elem) const;
+
+  ADReal getNeighborValue(const Elem * const neighbor,
+                          const FaceInfo & fi,
+                          const ADReal & elem_value) const;
+
+  ADReal
+  getFaceValue(const Elem * const neighbor, const FaceInfo & fi, const ADReal & elem_value) const;
+
+  const ADReal & getVertexValue(const Node & vertex) const;
```
:+1:

lindsayad replied:
I actually use `getElemValue` and `getNeighborValue` to build the Rhie-Chow interpolation in NSFV. I can make the others private though

lindsayad committed:
More this-> qualification in MooseVariableFV

lindsayad committed:
Selectively support same mat prop from different mats

We now allow different materials across an interface to define a
material property with the same name so long as the finite volume
physics are the same on either side of the interface

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/600359/) on 4dde2bd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15825/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15825/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bf084a9119b61fc6b2e75ab4fe62f1ca46d94218`

lindsayad commented:
Nice job iterating back and forth @rwcarlsen . With https://github.com/idaholab/moose/pull/15825/commits/7a1971dd06d4b14aeb6d22496db6e3bb6cc784c3 I'm quite happy that we can define only one variable and one kernel and do different diffusivities on different  subdomains

![Screenshot from 2020-10-21 15-53-39](https://user-images.githubusercontent.com/10248304/96798551-fbd54b00-13b5-11eb-8a16-e172afa94f09.png)

lindsayad commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -365,7 +414,103 @@ class MooseVariableFV : public MooseVariableField<OutputType>
+   * Dynamics"
+   */
+  const ADReal &
+  adCoeff(const Elem * elem, void * context, ADReal (*fn)(const Elem * const, void *)) const;
```
I actually think is bad design. These coefficients are really specific to the Rhie-Chow interpolation and I put them here only because I didn't quickly think of another way to store them in a single cache of data. I'm going to strip this out for now and if I really don't think of a better way to do this I can add this back in the NSFV PR (but I really should think of a better way)

lindsayad committed:
Better documentation of new MooseVariableFV data and methods

lindsayad committed:
Rip out questionably designed adCoeff code

rwcarlsen reviewed (dismissed):
This looks good.  The relaxation on the material block restriction error is a good step in the right direction.  Over time we can hopefully get this relaxed even more (#15991).

lindsayad commented:
#15338 is merged, and now this is rebased. Will need a re-approval

cpgr commented:
Just wanted to say that I have been using this branch for some work this week and everything is working well with different materials on different subdomains, so thanks a lot for all your effort!

lindsayad commented:
That's really nice to hear. Thank you very much for your feedback Chris!

On Mon, Oct 26, 2020 at 6:45 PM Chris Green <notifications@github.com>
wrote:

> Just wanted to say that I have been using this branch for some work this
> week and everything is working well with different materials on different
> subdomains, so thanks a lot for all your effort!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15825#issuecomment-716923100>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4A6PTOZLKLLG65EV2LSMYQ3NANCNFSM4RRE5E2A>
> .
>

cpgr commented:
One thing I just noticed is that I can't do value-based adaptivity. For example, if you use a `ValueRangeMarker` and only FV variables, you get an error like `_variable is null in marker. Are you using a finite volume variable?`. Spatial-based adaptivity like `BoxMarker` does work ok though. Hopefully this can be added soon (I can try but I have no idea where to start for this).

lindsayad commented:
Haha, well at least the error message is pretty spot-on!

On Mon, Oct 26, 2020 at 10:48 PM Chris Green <notifications@github.com>
wrote:

> One thing I just noticed is that I can't do value-based adaptivity. For
> example, if you use a ValueRangeMarker and only FV variables, you get an
> error like _variable is null in marker. Are you using a finite volume
> variable?. Spatial-based adaptivity like BoxMarker does work ok though.
> Hopefully this can be added soon (I can try but I have no idea where to
> start for this).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15825#issuecomment-716999861>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4GCIFRUN4E75UO5WNLSMZNJ3ANCNFSM4RRE5E2A>
> .
>

cpgr commented:
![4jz4cq](https://user-images.githubusercontent.com/7664823/97263476-2b34fd80-1877-11eb-88f2-9192df1415ad.jpg)

lindsayad committed:
Support FV variables in QuadraturePointMarker derivative classes

lindsayad commented:
@cpgr try your value-based markers now that I've added https://github.com/idaholab/moose/pull/15825/commits/9893a195bcfb73e720280a762e70a1a70903a3a4

cpgr commented:
Yep, I think it is working now. Thanks again!

rwcarlsen commented:
Looks like you've got a porous flow adaptivity test crashing.

lindsayad committed:
Restrict threading for porous flow adaptivity tri test

Per https://civet.inl.gov/job/604046/ we seem to occasionally get some
crashes with this test with threading. I can't reproduce this on my
machine but that doesn't mean the bug isn't there. It could be this only
happens on a very heavily loaded slow machine (which was the case for
the CIVET crash). Maybe @cpgr can run this through helgrind or something
if he ever has time

rwcarlsen commented:
It looks like a few griffin tests are crashing and magpie fails to compile.

lindsayad commented:
> It looks like a few griffin tests are crashing and magpie fails to compile.

The Griffin tests have been crashing ad infinitum. I will look into a fix for Magpie

lindsayad committed:
Help Magpie build

cpgr commented:
I'll take a look at that porous flow test to see if I can work out why it crashed

moosebuild commented:
Job [Modules debug](https://civet.inl.gov/job/604500/) on 87aedd7 : invalidated by @lindsayad

lindsayad commented:
Ok everything passing except Griffin
